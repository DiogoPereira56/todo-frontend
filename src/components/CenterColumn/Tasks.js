/* eslint-disable no-unused-vars */
import { useRef, useCallback, useState } from 'react';
import { PropTypes } from 'prop-types';
import { Task, CheckBox, Content } from './CenterColumn.styles';

const Tasks = ({
    tasks,
    setChangeLayout,
    setActiveTask,
    loggedIdClient,
    setPage,
    loading,
    hasMore,
    updateCompletion,
}) => {
    const show = (task) => {
        setActiveTask(task);
        setChangeLayout((prevLayout) => {
            return !prevLayout;
        });
    };

    const handleCompletedTask = (e, task) => {
        e.stopPropagation();
        const values = { idTask: task.idTask, complete: !task.complete, idClient: loggedIdClient };
        updateCompletion({ variables: values });
    };

    const observer = useRef();
    const lastTaskRef = useCallback(
        (node) => {
            if (loading) return;
            if (observer.current) observer.current.disconnect();
            observer.current = new IntersectionObserver((entries) => {
                if (entries[0].isIntersecting && hasMore > 0) {
                    //console.log('visible');
                    setPage((prevPage) => {
                        return prevPage + 1;
                    });
                }
            });
            if (node) observer.current.observe(node);
        },
        [loading, hasMore],
    );

    if (tasks) {
        return (
            <div>
                <Content data-testid="allTasks">
                    {tasks.map((task, index) => {
                        if (tasks.length == index + 1) {
                            return (
                                <Task
                                    data-testid="lastTask"
                                    ref={lastTaskRef}
                                    key={task.idTask}
                                    onClick={() => show(task)}
                                >
                                    <CheckBox
                                        data-testid="lastCheckBox"
                                        type="checkbox"
                                        defaultChecked={task.complete}
                                        onClick={(e) => handleCompletedTask(e, task)}
                                    />
                                    {task.title}
                                </Task>
                            );
                        } else {
                            return (
                                <Task key={task.idTask} onClick={() => show(task)}>
                                    <CheckBox
                                        type="checkbox"
                                        defaultChecked={task.complete}
                                        onClick={(e) => handleCompletedTask(e, task)}
                                    />
                                    {task.title}
                                </Task>
                            );
                        }
                    })}
                    {/* loading && <Task ref={lastTaskRef}>Loading...</Task> */}
                </Content>
            </div>
        );
    }

    return <div>error</div>;
};

Tasks.propTypes = {
    tasks: PropTypes.array,
    setChangeLayout: PropTypes.func,
    setActiveTask: PropTypes.func,
    loggedIdClient: PropTypes.number,
    setPage: PropTypes.func,
    loading: PropTypes.bool,
    hasMore: PropTypes.number,
    updateCompletion: PropTypes.func,
};

export default Tasks;
