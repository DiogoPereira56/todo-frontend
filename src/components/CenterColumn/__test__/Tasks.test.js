/* eslint-disable no-unused-vars */
/* eslint-disable no-undef */
import '../__test__/__mocks__/intersectionObserverMock';
import Tasks from '../Tasks';
import ReactDOM from 'react-dom';
import { fireEvent, prettyDOM, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';

const mockGetTasks = () => {
    return [
        { complete: false, description: '', idTask: 300, title: 't', __typename: 'task' },
        { complete: true, description: '', idTask: 301, title: 'j', __typename: 'task' },
    ];
};
/* const id = 3;
const loading = false;
const hasmore = -12; */

describe('<Tasks />', () => {
    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(<Tasks></Tasks>, div);
    });

    it('matches Tasks snapshot', () => {
        const tree = renderer.create(<Tasks></Tasks>).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders the last Task correctly', () => {
        const mockTasks = mockGetTasks();
        const { debug, getByTestId } = render(<Tasks tasks={mockTasks}></Tasks>);
        //debug();
        expect(getByTestId('lastTask')).toHaveTextContent('j');
    });

    it('clicks the Task correctly changing the layout', () => {
        const mockChangeLayout = jest.fn();
        const mockSetActiveTask = jest.fn();
        const mockTasks = mockGetTasks();
        const { debug, getByTestId } = render(
            <Tasks
                tasks={mockTasks}
                setChangeLayout={mockChangeLayout}
                setActiveTask={mockSetActiveTask}
            ></Tasks>,
        );
        //debug();
        fireEvent.click(getByTestId('lastTask'));
        expect(mockChangeLayout).toHaveBeenCalledTimes(1);
    });

    it('checks if the checkbox is updated correctly', () => {
        const mockUpdateCompletion = jest.fn();
        const mockTasks = mockGetTasks();
        const { debug, getByTestId } = render(
            <Tasks tasks={mockTasks} updateCompletion={mockUpdateCompletion}></Tasks>,
        );
        //debug();
        const checkbox = getByTestId('lastCheckBox');
        //console.log(prettyDOM(checkbox));
        expect(mockUpdateCompletion).toHaveBeenCalledTimes(0);
        fireEvent.click(checkbox);
        expect(mockUpdateCompletion).toHaveBeenCalledTimes(1);
    });
});
