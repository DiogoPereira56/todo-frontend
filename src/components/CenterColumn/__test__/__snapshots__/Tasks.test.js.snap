// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tasks /> matches Tasks snapshot 1`] = `
<div>
  error
</div>
`;
